Content.Server/Fluids/EntitySystems/PuddleSystem.Overflow.cs

```diff
@@ -0,0 +1,200 @@
+                }
+                puddleExpansion.PuddlesToExpand.Clear();
+
+                var max = nextToExpand.Count;
+
+                if (max == 0)
+                {
+
+                    continue;
+                }
+
+                var transferMax = FixedPoint2.Min(PuddleComponent.DefaultOverflowVolume,
+                    puddleExpansion.RemainingSolution.CurrentVolume / max);
+
+                foreach (var puddlePlacer in nextToExpand)
+                {
+                    if (puddleExpansion.RemainingSolution.CurrentVolume <= FixedPoint2.Zero)
+                        continue;
+
+                    var puddle = puddlePlacer.Invoke();
+                    puddleExpansion.PuddlesToExpand.Add(puddle);
+                    visitedEntities++;
+
+                    if (puddle.CurrentVolume >= PuddleComponent.DefaultOverflowVolume)
+                        continue;
+
+                    TryAddSolution(puddle.Owner, puddleExpansion.RemainingSolution.SplitSolution(transferMax), false, false,
+                        puddle);
+                }
+
+                if (visitedEntities < VisitedEntitiesLimit) continue;
+
+                _puddleExpansionQueue.Add(puddleExpansion);
+                return;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Find or get all 8 direction puddles that have less volume than the expandingPuddle
+    /// </summary>
+    /// <param name="sourcePuddle"></param>
+    /// <param name="puddleExpansion"></param>
+    /// <returns></returns>
+    private List<Func<PuddleComponent>> ExpandPuddles(PuddleComponent sourcePuddle, ref PuddleExpansion puddleExpansion)
+    {
+        MetaDataComponent? metadata = null;
+        TransformComponent? transform = null;
+
+        if (!Resolve(sourcePuddle.Owner, ref metadata, ref transform, false))
+            return new List<Func<PuddleComponent>>();
+
+        var mapGrid = _mapManager.GetGrid(transform.GridID);
+        var puddlesToExpand = new List<Func<PuddleComponent>>();
+        var tilePos = transform.Coordinates.ToVector2i(EntityManager, _mapManager);
+        var prototypeId = metadata.EntityPrototype!.ID;
+
+        puddleExpansion.VisitedTiles.Add(tilePos);
+        var directions = SharedDirectionExtensions.RandomDirections().ToArray();
+        foreach (var direction in directions)
+        {
+            var newPos = tilePos.Offset(direction);
+            if (puddleExpansion.VisitedTiles.Contains(newPos)) continue;
+
+            puddleExpansion.VisitedTiles.Add(newPos);
+            if (TryGetAdjacentPuddle(prototypeId, transform, direction, mapGrid, out var puddle))
+            {
+                puddlesToExpand.Add(puddle);
+            }
+        }
+
+
+        return puddlesToExpand;
+    }
+
+    private bool TryGetAdjacentPuddle(string prototype, TransformComponent originPuddle, Direction direction,
+        IMapGrid mapGrid, [NotNullWhen(true)] out Func<PuddleComponent>? puddlePlacer)
```

Comment:
Why does this return a function?

Reply:
Well:
1. It was like that before more or less (ease of comparing diffs)
2. Alternative is a struct that either creates a puddle or reuses the existing one. `Func` is cleaner in its intent, than `<SomeComponent>Placer`.